File: htmlParser/app/src/main/java/com/daeseong/htmlParser/MainActivity.java
package com.daeseong.htmlParser;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.graphics.Bitmap;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Button;
import android.widget.TextView;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private WebView webview;
    private TextView tv;
    private Button button1;

    private String s1;
    private String s2;
    private String s3;
    private String s4;
    private String s5;

    private String num1, num2, num3, num4, num5, num6, num7;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        webview = (WebView)findViewById(R.id.webview);
        tv = (TextView)findViewById(R.id.tv);
        tv.setMovementMethod(new ScrollingMovementMethod());

        WebSettings webSettings = webview.getSettings();
        webSettings.setJavaScriptEnabled(true);// 자바 스크립트 사용
        webview.addJavascriptInterface(new JavaScriptInterface(), "Android");

        webSettings.setSupportZoom(false);//확대 축소 기능
        webSettings.setCacheMode(WebSettings.LOAD_NO_CACHE);//캐시모드를 사용하지 않고 네트워크를 통해서만 호출
        webSettings.setAppCacheEnabled(false);//앱 내부 캐시 사용 여부 설정

        webSettings.setUseWideViewPort(true);//웹뷰에 맞게 출력하기
        webSettings.setLoadWithOverviewMode(true);
        webSettings.setBuiltInZoomControls(false); // 안드로이드 내장 줌 컨트롤 사용 X

        webview.setWebViewClient(new CustomWebViewClient());

        //네트워크 연결 여부
        if(IsConnect()){
            webview.loadUrl("http://www.lotto.co.kr/article/list/AC01");
        }else {
            webview.loadUrl("about:blank");
        }

        button1 = (Button)findViewById(R.id.button1);
        button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {


            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();

    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    private boolean IsConnect(){

        boolean bConnected = false;
        try {
            ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
            if (networkInfo != null && networkInfo.isConnected())
                bConnected = true;
        }catch (Exception e){
            e.printStackTrace();
        }
        return bConnected;
    }

    private class CustomWebViewClient extends WebViewClient {

        @Override
        public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
            super.onReceivedError(view, request, error);

        }

        @Override
        public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
            super.onReceivedError(view, errorCode, description, failingUrl);

        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {

            return super.shouldOverrideUrlLoading(view, request);
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {

            return super.shouldOverrideUrlLoading(view, url);
        }

        @Override
        public void onPageFinished(WebView view, String url) {

            view.loadUrl("javascript:window.Android.getHtml(document.getElementsByTagName('body')[0].innerHTML);");
            super.onPageFinished(view, url);
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            super.onPageStarted(view, url, favicon);
        }
    }

    public class JavaScriptInterface {

        @JavascriptInterface
        public void getHtml(String html) {

            try {

                getData(html);

                s3 = String.format("%s %s %s %s %s %s %s", num1, num2, num3, num4, num5, num6, num7);
                String sValue1 = String.format("회차: %s \n", s1);
                String sValue2 = String.format("추첨일: %s \n", s2);
                String sValue3 = String.format("당첨번호: %s \n", s3);
                String sValue4 = String.format("1등 당첨자수 : %s \n", s4);
                String sValue5 = String.format("1등 당첨금액 : %s \n", s5);
                tv.setText(sValue1 + sValue2 + sValue3 + sValue4 + sValue5);

            }catch (Exception ex){
                ex.getMessage().toString();
            }

        }

        private void getData(String html){

            Document document = Jsoup.parse(html);
            if (document != null) {

                Elements rowlist = document.select("ul").select("[class=wnr_cur_list]");
                if(rowlist != null){

                    Elements liList = rowlist.select("li");
                    if(liList != null){

                        Elements plist = liList.get(0).select("p");
                        if(plist != null){

                            Elements alist = plist.select("a");
                            if(alist != null){

                                int i=0;
                                Elements imglist = alist.select("img");
                                for(Element row : imglist){

                                    String src = row.attr("src");
                                    if( isNumeric(getOnlyFileName(src)) ){

                                        if(i == 0){
                                            num1 = getOnlyFileName(src);
                                        }else if(i == 1){
                                            num2 = getOnlyFileName(src);
                                        }else if(i == 2){
                                            num3 = getOnlyFileName(src);
                                        }else if(i == 3){
                                            num4 = getOnlyFileName(src);
                                        }else if(i == 4){
                                            num5 = getOnlyFileName(src);
                                        }else if(i == 5){
                                            num6 = getOnlyFileName(src);
                                        }else if(i == 6){
                                            num7 = getOnlyFileName(src);
                                        }
                                        i++;
                                    }
                                }
                            }

                            int i=0;
                            Elements spanlist = plist.select("span");
                            for(Element row : spanlist){

                                if(i == 0){
                                    s1 = row.text();
                                }else if(i == 1){
                                    s2 = row.text();
                                }else if(i == 2){
                                    s3 = row.text();
                                }else if(i == 3){
                                    s4 = row.text();
                                }else if(i == 4){
                                    s5 = row.text();
                                }
                                i++;
                            }

                        }
                    }

                }
            }
        }

        private String getOnlyFileName(String url){
            String sResult = "";
            int nIndex = url.lastIndexOf("/");
            if(nIndex >= 0){
                String sTemp = url.substring(nIndex +1 );
                nIndex = sTemp.lastIndexOf(".");
                if(nIndex >= 0){
                    sResult = sTemp.substring(0, nIndex);
                }
            }else {
                nIndex = url.lastIndexOf(".");
                if(nIndex >= 0){
                    sResult = url.substring(0, nIndex);
                }
            }
            return sResult;
        }

        //숫자만 정규식
        private boolean isNumeric(String sInput) {
            String regex = "[\\d]*$";
            return sInput.matches(regex);
        }
    }

}

File: htmlParser(kotlin)/app/src/main/java/com/daeseong/htmlparser/MainActivity.kt
package com.daeseong.htmlparser

import android.content.Context
import android.graphics.Bitmap
import android.net.ConnectivityManager
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.webkit.*
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import org.jsoup.Jsoup

class MainActivity : AppCompatActivity() {

    private val tag = MainActivity::class.java.simpleName

    private lateinit var webview: WebView
    private lateinit var tv: TextView
    private lateinit var button1: Button

    private var s1: String? = null
    private var s2: String? = null
    private var s3: String? = null
    private var s4: String? = null
    private var s5: String? = null

    private var num1: String? = null
    private var num2: String? = null
    private var num3: String? = null
    private var num4: String? = null
    private var num5: String? = null
    private var num6: String? = null
    private var num7: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        webview = findViewById(R.id.webview)
        tv = findViewById(R.id.tv)
        tv.movementMethod = ScrollingMovementMethod()

        val webSettings = webview.settings
        webSettings.javaScriptEnabled = true
        webview.addJavascriptInterface(JavaScriptInterface(), "Android")

        webSettings.setSupportZoom(false)//확대 축소 기능
        webSettings.cacheMode = WebSettings.LOAD_NO_CACHE
        webSettings.useWideViewPort = true
        webSettings.loadWithOverviewMode = true
        webSettings.builtInZoomControls = false

        webview.webViewClient = CustomWebViewClient()

        if (isConnect()) {
            webview.loadUrl("http://www.lotto.co.kr/article/list/AC01")
        } else {
            webview.loadUrl("about:blank")
        }

        button1 = findViewById(R.id.button1)
        button1.setOnClickListener {

        }
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onBackPressed() {
        super.onBackPressed()
    }

    private fun isConnect(): Boolean {
        val connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = connectivityManager.activeNetworkInfo ?: return false
        return networkInfo.isConnected && networkInfo.isAvailable
    }

    private fun getOnlyFileName(url: String): String? {
        var sResult = ""
        var nIndex = url.lastIndexOf("/")
        if (nIndex >= 0) {
            val sTemp = url.substring(nIndex + 1)
            nIndex = sTemp.lastIndexOf(".")
            if (nIndex >= 0) {
                sResult = sTemp.substring(0, nIndex)
            }
        } else {
            nIndex = url.lastIndexOf(".")
            if (nIndex >= 0) {
                sResult = url.substring(0, nIndex)
            }
        }
        return sResult
    }

    private fun isNumeric(sInput: String): Boolean {
        val regex = Regex("[\\d]*$")
        return sInput.matches(regex)
    }

    private inner class CustomWebViewClient : WebViewClient() {
        override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
            super.onReceivedError(view, request, error)
        }

        override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
            return super.shouldOverrideUrlLoading(view, request)
        }

        override fun shouldOverrideUrlLoading(view: WebView?, url: String): Boolean {
            return super.shouldOverrideUrlLoading(view, url)
        }

        override fun onPageFinished(view: WebView?, url: String?) {
            view?.loadUrl("javascript:window.Android.getHtml(document.getElementsByTagName('body')[0].innerHTML);")
            super.onPageFinished(view, url)
        }

        override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
            super.onPageStarted(view, url, favicon)
        }
    }

    private inner class JavaScriptInterface {
        @JavascriptInterface
        fun getHtml(html: String) {
            try {

                getData(html)

                s3 = String.format("%s %s %s %s %s %s %s", num1, num2, num3, num4, num5, num6, num7)
                val sValue1 = String.format("회차: %s \n", s1)
                val sValue2 = String.format("추첨일: %s \n", s2)
                val sValue3 = String.format("당첨번호: %s \n", s3)
                val sValue4 = String.format("1등 당첨자수 : %s \n", s4)
                val sValue5 = String.format("1등 당첨금액 : %s \n", s5)
                tv.text = sValue1 + sValue2 + sValue3 + sValue4 + sValue5

            } catch (ex: Exception) {
                ex.message.toString()
            }
        }

        private fun getData(html: String) {
            val document = Jsoup.parse(html)
            if (document != null) {
                val rowlist = document.select("ul").select("[class=wnr_cur_list]")
                if (rowlist != null) {
                    val liList = rowlist.select("li")
                    if (liList != null) {
                        val plist = liList[0].select("p")
                        if (plist != null) {
                            val alist = plist.select("a")
                            if (alist != null) {
                                var i = 0
                                val imglist = alist.select("img")
                                for (row in imglist) {
                                    val src = row.attr("src")
                                    if (isNumeric(getOnlyFileName(src)!!)) {
                                        when (i) {
                                            0 -> num1 = getOnlyFileName(src)
                                            1 -> num2 = getOnlyFileName(src)
                                            2 -> num3 = getOnlyFileName(src)
                                            3 -> num4 = getOnlyFileName(src)
                                            4 -> num5 = getOnlyFileName(src)
                                            5 -> num6 = getOnlyFileName(src)
                                            6 -> num7 = getOnlyFileName(src)
                                        }
                                        i++
                                    }
                                }
                            }

                            val spanlist = plist.select("span")
                            for ((i, row) in spanlist.withIndex()) {
                                when (i) {
                                    0 -> s1 = row.text()
                                    1 -> s2 = row.text()
                                    2 -> s3 = row.text()
                                    3 -> s4 = row.text()
                                    4 -> s5 = row.text()
                                }
                            }
                        }
                    }
                }
            }
        }

    }
}
